src/streaming/array.writer.cpp

```diff
@@ -0,0 +1,422 @@
+  , is_finalizing_{ false }
+{
+}
+
+size_t
+zarr::ArrayWriter::write_frame(const uint8_t* data, size_t nbytes)
+{
+    const size_t nbytes_frame =
+      bytes_of_frame(*config_.dimensions, config_.dtype);
+
+    if (nbytes_frame != nbytes) {
+        LOG_WARNING("Frame size mismatch: expected %zu, got %zu. Skipping",
+                    nbytes_frame,
+                    nbytes);
+        return 0;
+    }
+
+    if (chunk_buffers_.empty()) {
+        make_buffers_();
+    }
+
+    // split the incoming frame into tiles and write_frame them to the chunk
+    // buffers
+    const auto bytes_written = write_frame_to_chunks_(data, nbytes);
+    EXPECT(bytes_written == nbytes, "Failed to write_frame frame to chunks");
+
+    LOG_DEBUG("Wrote %zu bytes of frame %zu", bytes_written, frames_written_);
+    bytes_to_flush_ += bytes_written;
+    ++frames_written_;
+
+    if (should_flush_()) {
+        flush_();
+    }
+
+    return bytes_written;
+}
+
+bool
+zarr::ArrayWriter::make_data_sinks_()
+{
+    std::string data_root;
+    std::function<size_t(const ZarrDimension&)> parts_along_dimension;
+    switch (version_()) {
+        case ZarrVersion_2:
+            parts_along_dimension = chunks_along_dimension;
+            data_root = config_.store_path + "/" +
+                        std::to_string(config_.level_of_detail) + "/" +
+                        std::to_string(append_chunk_index_);
+            break;
+        case ZarrVersion_3:
+            parts_along_dimension = shards_along_dimension;
+            data_root = config_.store_path + "/data/root/" +
+                        std::to_string(config_.level_of_detail) + "/c" +
+                        std::to_string(append_chunk_index_);
+            break;
+        default:
+            LOG_ERROR("Unsupported Zarr version");
+            return false;
+    }
+
+    SinkCreator creator(thread_pool_, s3_connection_pool_);
+
+    if (config_.bucket_name && !creator.make_data_sinks(*config_.bucket_name,
+                                                        data_root,
+                                                        config_.dimensions,
+                                                        parts_along_dimension,
+                                                        data_sinks_)) {
+        LOG_ERROR("Failed to create data sinks in %s for bucket %s",
+                  data_root.c_str(),
+                  config_.bucket_name->c_str());
+        return false;
+    } else if (!config_.bucket_name &&
+               !creator.make_data_sinks(data_root,
+                                        config_.dimensions,
+                                        parts_along_dimension,
+                                        data_sinks_)) {
+        LOG_ERROR("Failed to create data sinks in %s", data_root.c_str());
+        return false;
+    }
+
+    return true;
+}
+
+bool
+zarr::ArrayWriter::make_metadata_sink_()
+{
+    if (metadata_sink_) {
+        return true;
```

Comment:
This logic is confusing. If a function named `make_metadata_sink_` returns true, I'd expect it to have created a new metadata sink. However, in this case, it returns true even when a new sink isn't created (if it was created previously). I suggest either logging an error and returning false, or at the very least adding an info log before returning true.